src/frontend/src/binder/select.rs

```diff
@@ -769,14 +770,43 @@ impl Binder {
             .unzip()
     }
 
-    fn bind_distinct_on(&mut self, distinct: Distinct) -> Result<BoundDistinct> {
+    /// Bind `DISTINCT` clause in a [`Select`].
+    ///
+    /// # Arguments
+    ///
+    /// * `name_to_index` - output column name -> index. Ambiguous (duplicate) output names are
+    ///   marked with `usize::MAX`.
+    fn bind_distinct_on(
+        &mut self,
+        distinct: Distinct,
+        name_to_index: &HashMap<String, usize>,
+        select_items: &[ExprImpl],
+    ) -> Result<BoundDistinct> {
         Ok(match distinct {
             Distinct::All => BoundDistinct::All,
             Distinct::Distinct => BoundDistinct::Distinct,
             Distinct::DistinctOn(exprs) => {
                 let mut bound_exprs = vec![];
                 for expr in exprs {
-                    bound_exprs.push(self.bind_expr(expr)?);
```

Comment:
I found another problem (not caused by this PR, but existing issue). The `expr` in `DISTINCT ON (expr)` is actually not arbitraty expression.

> The DISTINCT ON expressions are interpreted using the same rules as for ORDER BY (see above).

> Each expression can be the name or ordinal number of an output column (SELECT list item), or it can be an arbitrary expression formed from input-column values.

https://www.postgresql.org/docs/current/sql-select.html#SQL-ORDERBY

So in our current (including after this PR) implementation, these are problematic:
```
--- This has wrong meaning
SELECT DISTINCT ON(1) * from t;

--- This should be banned, but IMO doesn't have much hurt.
SELECT DISTINCT ON('a') * from t;

--- ditto
SELECT * from t ORDER BY 'a';

```

Reply:
I fixed this, could you PTAL?